core/src/main/scala/cromwell/core/simpleton/WomValueBuilder.scala

```diff
@@ -85,46 +99,95 @@ object WomValueBuilder {
         case MapElementPattern("right", more) => PairRight -> component.copy(path = more)
       }
     }
-
-    // Group tuples by key using a Map with key type `K`.
-    def group[K](tuples: Traversable[(K, SimpletonComponent)]): Map[K, Traversable[SimpletonComponent]] = {
-      tuples groupBy { case (i, _) => i } mapValues { _ map { case (_, s) => s} }
+    
+    def toWomFile(components: Traversable[SimpletonComponent]) = {
+      // If there's just one simpleton, it's a primitive (file or directory)
+      if (components.size == 1) components.asPrimitive
+      else {
+        // Otherwise make a map of the components and detect the type of file from the class field
+        val groupedListing = components.asMap
+        
+        def isClass(className: String) = {
+          groupedListing.get(ClassKey)
+          /* If the class field is in an array it will be prefixed with a ':', so check for that as well.
+           * e.g: secondaryFiles[0]:class -> "File" 
+           *      secondaryFiles[0]:value -> "file/path" 
+           * would produce a Map(
+           *  ":class" -> List(Simpleton("File")),
+           *  ":value" -> List(Simpleton("file/path"))
+           * )
+           */
+          .orElse(groupedListing.get(s":$ClassKey"))
+          .map(_.asPrimitive.valueString)
+          .contains(className)
+        } 
+          
+        def isDirectory = isClass(WomValueSimpleton.DirectoryClass)
+        def isFile = isClass(WomValueSimpleton.FileClass)
+
+        if (isDirectory) toWomValue(WomMaybeListedDirectoryType, components)
+        else if (isFile) toWomValue(WomMaybePopulatedFileType, components)
+        else throw new IllegalArgumentException(s"There is no WomFile that can be built from simpletons: ${groupedListing.toList.mkString(", ")}")
+      }
     }
-
+    
     outputType match {
       case _: WomPrimitiveType =>
-        components collectFirst { case SimpletonComponent(_, v) => v } get
+        components.asPrimitive
       case opt: WomOptionalType =>
         if (components.isEmpty) {
           WomOptionalValue(opt.memberType, None)
         } else {
           WomOptionalValue(toWomValue(opt.memberType, components))
         }
       case arrayType: WomArrayType =>
-        val groupedByArrayIndex: Map[Int, Traversable[SimpletonComponent]] = group(components map descendIntoArray)
-        WomArray(arrayType, groupedByArrayIndex.toList.sortBy(_._1) map { case (_, s) => toWomValue(arrayType.memberType, s) })
+        WomArray(arrayType, components.asArray map { toWomValue(arrayType.memberType, _) })
       case mapType: WomMapType =>
-        val groupedByMapKey: Map[String, Traversable[SimpletonComponent]] = group(components map descendIntoMap)
         // map keys are guaranteed by the WDL spec to be primitives, so the "coerceRawValue(..).get" is safe.
```

Comment:
What about the CWL spec?

Reply:
eg what if the keys are files, and the files are objects?

Reply:
Side note: Should have added ToL... I don't think we should worry about this right now

Reply:
Good point, and yeah I feel like we can punt on this case for now but I'll make a ticket if that's ok